### DISCUSSION TITLE
Logging thread busy loop in z_impl_log_process function

### DISCUSSION THREAD

**Discussion by JungoLin1978** [zephyrproject-rtos/zephyr#68289] [Q&A]
I am experiencing an issue with the behavior of the logging thread.
I want to verify whether this behavior is expected or not.

Below are the configurations on my platform.

The configuration in my platform is listed below:
#
```
#
# SMP Options
#
# CONFIG_SMP is not set
CONFIG_USE_SWITCH=y
CONFIG_USE_SWITCH_SUPPORTED=y
CONFIG_MP_NUM_CPUS=1
CONFIG_MP_MAX_NUM_CPUS=1
# end of SMP Options

#
# Logging
#
CONFIG_LOG=y
CONFIG_LOG_CORE_INIT_PRIORITY=0
CONFIG_LOG_MODE_DEFERRED=y

#
# Processing
#
# CONFIG_LOG_PRINTK is not set
CONFIG_LOG_MODE_OVERFLOW=y
# CONFIG_LOG_BLOCK_IN_THREAD is not set
CONFIG_LOG_PROCESS_TRIGGER_THRESHOLD=10
CONFIG_LOG_PROCESS_THREAD=y
CONFIG_LOG_PROCESS_THREAD_STARTUP_DELAY_MS=0
CONFIG_LOG_PROCESS_THREAD_SLEEP_MS=100
CONFIG_LOG_PROCESS_THREAD_CUSTOM_PRIORITY=y
CONFIG_LOG_PROCESS_THREAD_PRIORITY=7
CONFIG_LOG_BUFFER_SIZE=65536
CONFIG_LOG_TRACE_SHORT_TIMESTAMP=y
# CONFIG_LOG_TIMESTAMP_64BIT is not set
CONFIG_LOG_SPEED=y
# end of Processing
# CONFIG_LOG_MODE_IMMEDIATE is not set
# CONFIG_LOG_MODE_MINIMAL is not set
# CONFIG_LOG_FRONTEND is not set
# CONFIG_LOG_CUSTOM_HEADER is not set
# CONFIG_LOG_MULTIDOMAIN is not set
``` 

When the issue is happened, the logging thread will be kept running z_impl_log_process(void) function because of z_log_msg_pending() always return true.
https://elixir.bootlin.com/zephyr/latest/source/subsys/logging/log_core.c#L544

We enable the MPSC_PBUF_DEBUG for this issue
https://elixir.bootlin.com/zephyr/v3.5.0/source/lib/os/mpsc_pbuf.c#L8

Here is the issue log (PS. some UART logs are lost)
```
11:19:40.333] alloc 10 words, wr:15756/15756, rd:15608/15608   ==> (Thread A)
11:19:40.333] allocated 0x810e6
                      alloc 22 words, wr:15756/15766, rd:15608/15608  ==> (Thread B)
11:19:40.333] allocated 0x810e6f88, wr:15756/15788, rd:15608/15608
11:19:40.360] >>claimed 158: 0x810e6ef8, wr:15888/15898, rd:15730/15756 ==> (Logging thread)
11:19:40.360] <<freed: 0x810e6ef8, wr:15888/15898, rd:15756/15756
11:19:40.360] invalid claim 132: 0x810e6f60, wr:15888/15898, rd:15756/15756
11:19:40.360] >>claimed 132: (nil), wr:15888/15898, rd:15756/15756
...
11:30:24.564] >>claimed 628: (nil), wr:172/182, rd:15756/15756
11:30:24.564] invalid claim 628: 0x810e6f60, wr:172/182, rd:15756/15756
``` 

The logging thread cannot claim the valid message item with buffer->tmp_rd_idx: 15756.
This message item with tmp_wr_idx:15756 was allocated with 10 words at 11:19:40.333 by thread A, but it was not committed due to a context switch to thread B. Furthermore, the priority of thread A is lower than that of threads B and the logging thread. As a result, there is always a 10-word offset between wr_idx and tmp_wr_idx in the MPSC_PBUF_DEBUG log. 
When the logging thread is active, it uses mpsc_pbuf_is_pending(struct mpsc_pbuf_buffer *buffer) to determine whether to release the CPU or not. However, the consistently true returned result leads to thread A experiencing starvation.
https://elixir.bootlin.com/zephyr/v3.5.0/source/lib/os/mpsc_pbuf.c#L606

My question is: 
a) Is this an issue? If not, where should I fix it? 
b) If yes, does it get fixed in the mpsc_pbuf_is_pending function to check the available buffer and the valid item at the same time?

I appreciate your feedback in advance. 

Best regards, 

Jungo

**Comment by lromeraj** [comment]
Hi @JungoLin1978, you may be interested in #101401. Yes, this is definitely an issue. There are multiple ways to address it: one is the approach you already mentioned, and another is the one I proposed in #101402, which should mitigate the problem until we settle on the best long-term solution.
